### DISCUSSION TITLE
RFC: Blocking revalidation

### DISCUSSION THREAD

**Discussion by andreaf96** [vercel/next.js#26058] [Help]
### Describe the feature you'd like to request

At the time I'm writing we can use {"fallback":"blocking"} in getStaticPaths() to make the client wait for the generation of a "new" page. We can also use {"revalidate":10} in getStaticProps() to make the server trigger a background regeneration of the page after someone visits the page after 10 seconds. The problem is that someone could have obscure pages that are almost never visited that may get way too out of sync with production data. 
It would be extremely useful if we could decide make the regeneration triggered by the "revalidate" timer be "blocking" just like when the user visits a new path that was not previously built. 

### Describe the solution you'd like

We could return an object with properties {"revalidate":10, "revalidate-strategy": "blocking"} to tell the server to achieve the result I asked above.

### Describe alternatives you've considered

Another option is to just make the behaviour I asked for be the new default when the user uses both "fallback":"blocking" in getStaticPaths() and "revalidate":10 in getStaticProps().

**Comment by lucasbasquerotto** [comment]
This was an option I'm expecting to have, because if a page is not accessed frequently, it could have very old stuff, which would be bad.

# Proposal

Add a purge property that can be used together with revalidate. The purge option would define the number of seconds to block the revalidation of a page.

```ts
export const getStaticProps: GetStaticProps = async ({ params }) => {
    // const props = ...
    return { props, revalidate: 60, purge: 600 };
};
```

In the above example:

1. If the page was generated less than 1 minute ago, the static page would be returned and nothing else would be done.
2. If the page was generated more than 1 minute ago but less than 10 minutes ago, the static page would be returned and a new static page would be generated in the background.
3. If the page was generated more than 10 minutes ago, a new (blocking) static page would be generated and returned.

This way, someone that access a page that was generated more than 3 days ago would see an up-to-date page, instead of a very old page.

**Reply by Nexum** [reply]
bump, i want this :D gives me a headache working with cms data, we need to get the data up to date on demand :( either this or a cache revalidation on demand

**Comment by leerob** [comment]
I'm incredibly excited to share that Next.js 12.1 was just released, which includes **on-demand Incremental Static Regeneration**! We know many of you have been looking forward to this for a while, including myself, and I'm so happy it is here. 

- [Learn about Next.js 12.1](https://nextjs.org/blog/next-12-1)
- [Watch a demo of on-demand ISR](https://youtu.be/BGexHR1tuOA)
- [View another demo and give us your feedback](https://on-demand-isr.vercel.app/)
- [Read the documentation](https://nextjs.org/docs/basic-features/data-fetching/incremental-static-regeneration#on-demand-revalidation-beta)

**Reply by lucasbasquerotto** [reply]
Hi, @leerob! Thanks, that is a great feature that a lot of people was waiting. Unfortunately, it doesn't seem to be the answer to this RFC, because it requires manual intervention for each page separatelly in an outside request (which is great for cache invalidation in a blog, for example). 

**This request is about all pages blocking the request and returning an up-to-date page when revalidate expires** (it would work as a `fallback: 'blocking'` behaves the 1st time the page is loaded, returning an up-to-date page, but in this case it wouldn't be the 1st time only, but **every time the revalidate expires**). 

**The objective of this request is to avoid pages that are rarely accessed from returning very old stuff.** That said, now that the on-demand ISR is implemented, maybe it should be very straightforward to implement this feature (I don't know the code internally tough, so I'm not sure).

**Reply by leerob** [reply]
If you need that approach, you can use SSR (`getServerSideProps`) + `cache-control` headers 🙏

**Reply by lucasbasquerotto** [reply]
Wouldn't that make each request from a different user cause a new render in the server side? Unless I use a reverse proxy (like nginx) in the middle to cache pages, but it wouldn't work with rewrites (and we use them a lot) because the same url can actually go to different pages depending on some factors (mainly, depending if the user is authenticated or not, returning a skeleton if the user is authenticated and loading the data from the client side). 

Caching in the reverse proxy would show unauthenticated pages to authenticated users and vice-versa. Otherwise, I would need generate a new page for each different user accessing our website and cache in the browser (and the same problem could happen if the user logged in), which isn't desired, especially because it would not change in most cases. BTW, all our pages currently use SSG.

**Reply by facundocapua** [reply]
It would be great to have something like @lucasbasquerotto suggested.
